ksm.go

```diff
@@ -0,0 +1,497 @@
+	}
+
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+type ksm struct {
+	root          string
+	run           sysfsAttribute
+	pagesToScan   sysfsAttribute
+	sleepInterval sysfsAttribute
+	initialized   bool
+
+	initialPagesToScan   string
+	initialSleepInterval string
+	initialKSMRun        string
+
+	currentKnob ksmMode
+	throttling  bool
+	kickChannel chan bool
+
+	sync.Mutex
+}
+
+func (k *ksm) isAvailable() error {
+	info, err := os.Stat(k.root)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", k.root)
+	}
+
+	return nil
+}
+
+func (k *ksm) init(root string) error {
+	var err error
+
+	k.Lock()
+	defer k.Unlock()
+
+	k.initialized = false
+	k.throttling = false
+	k.root = root
+
+	if root == "" {
+		return errors.New("Invalid KSM root")
+	}
+
+	if err := k.isAvailable(); err != nil {
+		return err
+	}
+
+	k.pagesToScan = sysfsAttribute{
+		path: filepath.Join(k.root, ksmPagesToScan),
+	}
+
+	k.sleepInterval = sysfsAttribute{
+		path: filepath.Join(k.root, ksmSleepMillisec),
+	}
+
+	k.run = sysfsAttribute{
+		path: filepath.Join(k.root, ksmRunFile),
+	}
+
+	if err := k.run.open(); err != nil {
+		return err
+	}
+
+	if err := k.sleepInterval.open(); err != nil {
+		_ = k.run.close()
+		return err
+	}
+
+	if err := k.pagesToScan.open(); err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialPagesToScan, err = k.pagesToScan.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialSleepInterval, err = k.sleepInterval.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialKSMRun, err = k.run.read()
+	if err != nil {
+		_ = k.run.close()
+		_ = k.sleepInterval.close()
+		return err
+	}
+
+	k.initialized = true
+	k.kickChannel = make(chan bool)
+
+	return nil
+}
+
+// restoreSysFS is unlocked. You should take the ksm lock before calling it.
+func (k *ksm) restoreSysFS() error {
+	var err error
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	if err = k.pagesToScan.write(k.initialPagesToScan); err != nil {
+		return err
+	}
+
+	if err = k.sleepInterval.write(k.initialSleepInterval); err != nil {
+		return err
+	}
+
+	if err = k.run.write(k.initialKSMRun); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (k *ksm) restore() error {
+	var err error
+
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	if err = k.restoreSysFS(); err != nil {
+		return err
+	}
+
+	if err := k.run.close(); err != nil {
+		return err
+	}
+
+	if err := k.sleepInterval.close(); err != nil {
+		return err
+	}
+
+	if err := k.pagesToScan.close(); err != nil {
+		return err
+	}
+
+	k.initialized = false
+	return nil
+}
+
+func (k *ksm) tune(s ksmSetting) error {
+	k.Lock()
+	defer k.Unlock()
+
+	if !k.initialized {
+		return errors.New("KSM is unavailable")
+	}
+
+	if !s.run {
+		return k.run.write(ksmStop)
```

Comment:
ooh - sneaky short circuit return with side effect (turn off ksm) - took me a little to spot that ;-)